File: frontend/src/stores/messages.ts
import { computed, ref, watch } from 'vue'
import { defineStore } from 'pinia'
import type { RealtimeChannel } from '@supabase/supabase-js'
import { api } from '@/lib/api'
import { supabase } from '@/lib/supabase'
import { mockMessagesByThread, mockThreads } from '@/mocks/messages'
import { useNotificationsStore } from '@/stores/notifications'

export interface MessageThread {
  id: string
  participantId: string
  participantDisplayName: string
  lastMessagePreview: string | null
  updatedAt: string
  unreadCount: number
  muted: boolean
}

export interface ChatMessage {
  id: string
  threadId: string
  senderId: string
  body: string
  imageUrl: string | null
  createdAt: string
}

/** Raw `public.messages` row shape as Supabase Realtime delivers it (snake_case DB columns,
 * not the backend's camelCase `ChatMessage`). */
interface MessageRow {
  id: string
  thread_id: string
  sender_id: string
  body: string
  image_url: string | null
  created_at: string
}

/** Messages a conversation opens with, and the size of each "Load earlier" page. Matches the
 * backend's own `_MESSAGE_PAGE_SIZE` default - sent explicitly so the two can't drift apart
 * without `hasMoreByThread` noticing. */
const PAGE_SIZE = 40

export const useMessagesStore = defineStore('messages', () => {
  const threads = ref<MessageThread[]>([])
  const threadsLoading = ref(false)
  const threadsError = ref<string | null>(null)

  const activeThreadId = ref<string | null>(null)
  const messagesByThread = ref<Record<string, ChatMessage[]>>({})
  const messagesLoading = ref(false)
  const messagesError = ref<string | null>(null)
  /** Whether an older page exists for a thread - a full page came back, so there may be more. */
  const hasMoreByThread = ref<Record<string, boolean>>({})
  const loadingEarlier = ref(false)

  const activeThread = computed(
    () => threads.value.find((t) => t.id === activeThreadId.value) ?? null,
  )
  const activeMessages = computed(() =>
    activeThreadId.value ? (messagesByThread.value[activeThreadId.value] ?? []) : [],
  )
  const activeHasMore = computed(() =>
    activeThreadId.value ? (hasMoreByThread.value[activeThreadId.value] ?? false) : false,
  )

  function appendMessage(threadId: string, message: ChatMessage) {
    const existing = (messagesByThread.value[threadId] ??= [])
    if (existing.some((m) => m.id === message.id)) return
    existing.push(message)
    const thread = threads.value.find((t) => t.id === threadId)
    if (thread) {
      // An image-only message has an empty body, same as the backend's own `_preview`.
      thread.lastMessagePreview = message.body || (message.imageUrl ? 'Photo' : '')
      thread.updatedAt = message.createdAt
    }
  }

  // Live updates for the open thread (CHECKPOINT.md 3.5b's RLS policies authorize each event
  // against the viewer's own JWT), replacing polling - subscribes/unsubscribes as
  // `activeThreadId` changes rather than staying on one thread's channel forever.
  let realtimeChannel: RealtimeChannel | null = null

  function unsubscribeRealtime() {
    if (realtimeChannel) {
      supabase.removeChannel(realtimeChannel)
      realtimeChannel = null
    }
  }

  function subscribeRealtime(threadId: string) {
    unsubscribeRealtime()
    realtimeChannel = supabase
      .channel(`messages:${threadId}`)
      .on<MessageRow>(
        'postgres_changes',
        {
          event: 'INSERT',
          schema: 'public',
          table: 'messages',
          filter: `thread_id=eq.${threadId}`,
        },
        ({ new: row }) => {
          appendMessage(threadId, {
            id: row.id,
            threadId: row.thread_id,
            senderId: row.sender_id,
            body: row.body,
            imageUrl: row.image_url,
            createdAt: row.created_at,
          })
        },
      )
      .subscribe()
  }

  watch(activeThreadId, (id) => {
    if (id) subscribeRealtime(id)
    else unsubscribeRealtime()
  })

  /** Messages page's thread list. Falls back to `mockThreads` if the request fails (signed out,
   * network error, backend down), same resilience convention as 3.1j/3.2d/3.4c. */
  async function fetchThreads() {
    threadsLoading.value = true
    threadsError.value = null
    try {
      const fetched = await api.get<MessageThread[]>('/messages/threads')
      // `MessagesPanel` fires this alongside `?with=`'s `startThread`, so a thread created
      // after the server had already read the list would otherwise vanish from the inbox the
      // moment this response lands - and take the open conversation down with it.
      const active = threads.value.find((t) => t.id === activeThreadId.value)
      threads.value =
        active && !fetched.some((t) => t.id === active.id) ? [active, ...fetched] : fetched
    } catch (err) {
      threadsError.value = err instanceof Error ? err.message : 'Failed to load chats'
      threads.value = [...mockThreads]
    } finally {
      threadsLoading.value = false
    }
  }

  /** Opening a thread - also marks the other party's messages read server-side (replaces the
   * old frontend-only `unreadCount = 0`), so this always refetches rather than reusing a cached
   * list even when messages for this thread were already loaded once. Only the newest
   * `PAGE_SIZE` come back; `loadEarlier` walks backwards from there (4.49). */
  async function fetchMessages(threadId: string) {
    // A cached transcript stays on screen while the refresh runs - re-opening a chat you were
    // just in shouldn't blank the pane and then repaint it.
    messagesLoading.value = !messagesByThread.value[threadId]?.length
    messagesError.value = null
    try {
      const page = await api.get<ChatMessage[]>(
        `/messages/threads/${threadId}/messages?limit=${PAGE_SIZE}`,
      )
      messagesByThread.value[threadId] = page
      hasMoreByThread.value[threadId] = page.length === PAGE_SIZE
      const thread = threads.value.find((t) => t.id === threadId)
      if (thread) thread.unreadCount = 0
    } catch (err) {
      messagesError.value = err instanceof Error ? err.message : 'Failed to load messages'
      messagesByThread.value[threadId] = mockMessagesByThread[threadId] ?? []
      hasMoreByThread.value[threadId] = false
    } finally {
      messagesLoading.value = false
    }
  }

  /** Prepends the page of messages older than the oldest one held for `threadId`. Errors are
   * swallowed into `hasMore` staying true rather than replacing the transcript with an error
   * state - the conversation on screen is still perfectly readable. */
  async function loadEarlier(threadId: string | null = activeThreadId.value) {
    if (!threadId || loadingEarlier.value || !hasMoreByThread.value[threadId]) return
    const existing = messagesByThread.value[threadId] ?? []
    const oldest = existing[0]
    if (!oldest) return
    loadingEarlier.value = true
    try {
      const page = await api.get<ChatMessage[]>(
        `/messages/threads/${threadId}/messages?limit=${PAGE_SIZE}&before=${encodeURIComponent(oldest.createdAt)}`,
      )
      const known = new Set(existing.map((m) => m.id))
      messagesByThread.value[threadId] = [...page.filter((m) => !known.has(m.id)), ...existing]
      hasMoreByThread.value[threadId] = page.length === PAGE_SIZE
    } finally {
      loadingEarlier.value = false
    }
  }

  async function selectThread(id: string) {
    activeThreadId.value = id
    await fetchMessages(id)
    // Chat alerts live on the header's messages badge rather than the bell, so reading the
    // conversation is what takes them down. Strictly best-effort and deliberately not awaited:
    // clearing a badge must never be the reason opening (or starting) a chat fails.
    try {
      void useNotificationsStore()
        .markThreadRead(id)
        .catch(() => {})
    } catch {
      // ignore
    }
  }

  /** Find-or-create a thread with `participantId` and select it (backend dedups regardless of
   * who initiates, CHECKPOINT.md 3.5a). */
  async function startThread(participantId: string) {
    const thread = await api.post<MessageThread>('/messages/threads', { participantId })
    const index = threads.value.findIndex((t) => t.id === thread.id)
    if (index === -1) threads.value.unshift(thread)
    else threads.value[index] = thread
    await selectThread(thread.id)
    return thread
  }

  /** Multipart rather than JSON (4.49) so an attached image rides along as a real file, the
   * same way the feed composer posts one - the caller compresses it first (`utils/image.ts`)
   * and the backend re-encodes it to WebP. Either the text or the image may be missing, not
   * both. */
  async function sendMessage(payload: { body?: string; image?: File | null }) {
    const threadId = activeThreadId.value
    const text = (payload.body ?? '').trim()
    if (!threadId || (!text && !payload.image)) return
    const formData = new FormData()
    formData.append('body', text)
    if (payload.image) formData.append('image', payload.image)
    const message = await api.post<ChatMessage>(`/messages/threads/${threadId}/messages`, formData)
    appendMessage(threadId, message)
    return message
  }

  /** Optimistic: flips `muted` immediately rather than waiting on the round trip, since nothing
   * about the mute state depends on the server's response - rolls back on failure. */
  async function muteThread(threadId: string, muted: boolean) {
    const thread = threads.value.find((t) => t.id === threadId)
    if (!thread) return
    const previous = thread.muted
    thread.muted = muted
    try {
      await api.patch(`/messages/threads/${threadId}/mute`, { muted })
    } catch (err) {
      thread.muted = previous
      throw err
    }
  }

  /** Optimistic removal, same rationale as `muteThread` - restores the thread (and active
   * selection) in place on failure. */
  async function deleteThread(threadId: string) {
    const index = threads.value.findIndex((t) => t.id === threadId)
    if (index === -1) return
    const [removed] = threads.value.splice(index, 1)
    const wasActive = activeThreadId.value === threadId
    if (wasActive) activeThreadId.value = null

    try {
      await api.delete(`/messages/threads/${threadId}`)
    } catch (err) {
      threads.value.splice(index, 0, removed)
      if (wasActive) activeThreadId.value = threadId
      throw err
    }

    if (wasActive) {
      delete messagesByThread.value[threadId]
      const fallbackId = threads.value[0]?.id
      if (fallbackId) await selectThread(fallbackId)
    }
  }

  return {
    threads,
    threadsLoading,
    threadsError,
    activeThreadId,
    activeThread,
    messagesByThread,
    activeMessages,
    messagesLoading,
    messagesError,
    hasMoreByThread,
    activeHasMore,
    loadingEarlier,
    fetchThreads,
    selectThread,
    loadEarlier,
    startThread,
    sendMessage,
    muteThread,
    deleteThread,
  }
})

File: backend/src/backend/routers/messages.py
from datetime import UTC, datetime
from uuid import uuid4

from fastapi import (
    APIRouter,
    BackgroundTasks,
    Depends,
    File,
    Form,
    HTTPException,
    Query,
    UploadFile,
    status,
)

from ..core.auth import get_current_user_id
from ..core.blocks import require_not_blocked
from ..core.notify import notify
from ..core.schema import CamelModel
from ..core.storage import upload_image_as_webp
from ..core.supabase import get_supabase_client

router = APIRouter(prefix="/messages", tags=["messages"])

#: How many messages a conversation opens with. Older ones come back page by page through
#: `before`, so opening a long-running chat costs one bounded read instead of the whole history.
_MESSAGE_PAGE_SIZE = 40
_MAX_MESSAGE_PAGE_SIZE = 100


# Schemas ---------------------------------------------------------------------------------


class StartThreadIn(CamelModel):
    participant_id: str


class MuteThreadIn(CamelModel):
    muted: bool


class MuteThreadOut(CamelModel):
    id: str
    muted: bool


class ThreadOut(CamelModel):
    id: str
    participant_id: str
    participant_display_name: str
    last_message_preview: str | None
    updated_at: str
    unread_count: int
    muted: bool


class MessageOut(CamelModel):
    id: str
    thread_id: str
    sender_id: str
    body: str
    image_url: str | None = None
    created_at: str


# Helpers -----------------------------------------------------------------------------------

_THREAD_SELECT = (
    "id, user_a_id, user_b_id, created_at, "
    "user_a:users!message_threads_user_a_id_fkey(display_name), "
    "user_b:users!message_threads_user_b_id_fkey(display_name)"
)


def _canonical_pair(user_id: str, participant_id: str) -> tuple[str, str]:
    """`message_threads` has one row per unordered pair (CHECKPOINT.md's Messages note), but its
    unique constraint is on the ordered `(user_a_id, user_b_id)` tuple - sorting the ids keeps
    every thread between the same two people stored (and found) under one canonical row
    regardless of who started it."""
    return (user_id, participant_id) if user_id < participant_id else (participant_id, user_id)


def _get_thread(client, thread_id: str, user_id: str) -> dict:
    result = client.table("message_threads").select(_THREAD_SELECT).eq("id", thread_id).maybe_single().execute()
    if not result or not result.data:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Thread not found")
    thread = result.data
    if user_id not in (thread["user_a_id"], thread["user_b_id"]):
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Thread not found")
    return thread


def _preview(body: str | None, image_url: str | None) -> str | None:
    """An image-only message stores an empty `body` (20260920090000), so the inbox line needs
    something to show for it."""
    if body:
        return body
    return "Photo" if image_url else None


def _thread_out(
    thread: dict, user_id: str, *, last_message: dict | None, unread_count: int, muted: bool = False
) -> dict:
    is_a = thread["user_a_id"] == user_id
    participant_id = thread["user_b_id"] if is_a else thread["user_a_id"]
    participant = (thread.get("user_b") if is_a else thread.get("user_a")) or {}
    return {
        "id": thread["id"],
        "participant_id": participant_id,
        "participant_display_name": participant.get("display_name") or "SquadUp user",
        "last_message_preview": _preview(
            last_message["body"] if last_message else None,
            (last_message or {}).get("image_url"),
        ),
        "updated_at": last_message["created_at"] if last_message else thread["created_at"],
        "unread_count": unread_count,
        "muted": muted,
    }


def _get_thread_state(client, thread_id: str, user_id: str) -> dict:
    result = (
        client.table("message_thread_states")
        .select("muted, deleted_at")
        .eq("thread_id", thread_id)
        .eq("user_id", user_id)
        .maybe_single()
        .execute()
    )
    return (result.data if result else None) or {"muted": False, "deleted_at": None}


def _set_thread_state(client, thread_id: str, user_id: str, **fields) -> None:
    client.table("message_thread_states").upsert(
        {"thread_id": thread_id, "user_id": user_id, **fields},
        on_conflict="thread_id,user_id",
    ).execute()


def _require_membership(client, thread_id: str, user_id: str) -> None:
    """Cheaper than `_get_thread` for endpoints that only need the auth check, not the joined
    display names - mute/delete don't return a full `ThreadOut`, so there's no reason to pay for
    that join."""
    result = (
        client.table("message_threads")
        .select("user_a_id, user_b_id")
        .eq("id", thread_id)
        .maybe_single()
        .execute()
    )
    if not result or not result.data:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Thread not found")
    if user_id not in (result.data["user_a_id"], result.data["user_b_id"]):
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Thread not found")


# Routes --------------------------------------------------------------------------------------


@router.get("/threads", response_model=list[ThreadOut])
def list_threads(user_id: str = Depends(get_current_user_id)) -> list[dict]:
    """Messages page's thread list. One inbox per account (CHECKPOINT.md's Messages note), so
    this covers both a plain user's and a Pal's view with no persona split.

    The whole list - participants, last message, unread tallies, mute flags, and the deleted and
    blocked filtering - comes back from `message_thread_summaries` in one round-trip (4.49). It
    used to be four, the last of which read every message in every one of the viewer's threads
    just to keep the newest of each.
    """
    client = get_supabase_client()
    rows = client.rpc("message_thread_summaries", {"p_user_id": user_id}).execute().data or []
    return [
        {
            "id": row["id"],
            "participant_id": row["participant_id"],
            "participant_display_name": row["participant_display_name"] or "SquadUp user",
            "last_message_preview": _preview(row["last_message_body"], row["last_message_image_url"]),
            "updated_at": row["updated_at"],
            "unread_count": row["unread_count"],
            "muted": row["muted"],
        }
        for row in rows
    ]


@router.post("/threads", response_model=ThreadOut, status_code=status.HTTP_201_CREATED)
def start_thread(payload: StartThreadIn, user_id: str = Depends(get_current_user_id)) -> dict:
    if payload.participant_id == user_id:
        raise HTTPException(status.HTTP_400_BAD_REQUEST, "Cannot start a thread with yourself")

    client = get_supabase_client()
    participant = client.table("users").select("id").eq("id", payload.participant_id).maybe_single().execute()
    if not participant or not participant.data:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "User not found")
    require_not_blocked(user_id, payload.participant_id, "message")

    user_a_id, user_b_id = _canonical_pair(user_id, payload.participant_id)
    existing = (
        client.table("message_threads")
        .select(_THREAD_SELECT)
        .eq("user_a_id", user_a_id)
        .eq("user_b_id", user_b_id)
        .maybe_single()
        .execute()
    )
    if existing and existing.data:
        thread = existing.data
        # Starting a thread you'd previously deleted brings it back into your list rather than
        # leaving it permanently hidden - matches `_get_thread_state`'s "no row = not deleted".
        state = _get_thread_state(client, thread["id"], user_id)
        if state["deleted_at"]:
            _set_thread_state(client, thread["id"], user_id, deleted_at=None)
    else:
        created = (
            client.table("message_threads")
            .insert({"user_a_id": user_a_id, "user_b_id": user_b_id})
            .execute()
        )
        thread = _get_thread(client, created.data[0]["id"], user_id)
        state = {"muted": False}

    return _thread_out(thread, user_id, last_message=None, unread_count=0, muted=state["muted"])


@router.get("/threads/{thread_id}/messages", response_model=list[MessageOut])
def list_messages(
    thread_id: str,
    background: BackgroundTasks,
    limit: int = Query(_MESSAGE_PAGE_SIZE, ge=1, le=_MAX_MESSAGE_PAGE_SIZE),
    before: str | None = Query(None, description="ISO timestamp - returns messages older than this"),
    user_id: str = Depends(get_current_user_id),
) -> list[dict]:
    """A page of the conversation, newest `limit` first from the database but returned oldest
    first so the caller can render it straight down the transcript. `before` walks backwards
    through the history for "Load earlier messages" (4.49) - the endpoint used to return every
    message in the thread, so a long chat paid for all of it on every open.

    Opening a thread also marks the other party's unread messages read, replacing the
    frontend-only `unreadCount = 0` `stores/messages.ts` previously did on `selectThread`. That
    UPDATE is nothing the response depends on, so it runs after the response is sent rather than
    holding the transcript behind a write.
    """
    client = get_supabase_client()
    _require_membership(client, thread_id, user_id)

    query = (
        client.table("messages")
        .select("id, thread_id, sender_id, body, image_url, created_at")
        .eq("thread_id", thread_id)
        .order("created_at", desc=True)
        .limit(limit)
    )
    if before:
        query = query.lt("created_at", before)
    rows = query.execute().data or []

    if not before:
        background.add_task(_mark_thread_read, thread_id, user_id)

    rows.reverse()
    return rows


def _mark_thread_read(thread_id: str, user_id: str) -> None:
    get_supabase_client().table("messages").update({"read_at": datetime.now(UTC).isoformat()}).eq(
        "thread_id", thread_id
    ).neq("sender_id", user_id).is_("read_at", "null").execute()


@router.post(
    "/threads/{thread_id}/messages",
    response_model=MessageOut,
    status_code=status.HTTP_201_CREATED,
)
def send_message(
    thread_id: str,
    body: str = Form(""),
    image: UploadFile | None = File(None),  # noqa: B008
    user_id: str = Depends(get_current_user_id),
) -> dict:
    """Multipart rather than JSON (4.49) so an attached image rides along as a real file, same
    shape as the feed composer's `POST /feed/posts`. It is re-encoded to WebP and downscaled by
    `upload_image_as_webp` before it lands in the `message-images` bucket, on top of the
    downscale the client already does - a multi-MB phone photo has no business being stored at
    full size for a chat bubble. Either the text or the image may be missing, not both."""
    body = body.strip()
    if not body and image is None:
        raise HTTPException(status.HTTP_400_BAD_REQUEST, "Message must have text or an image")

    client = get_supabase_client()
    thread = _get_thread(client, thread_id, user_id)
    other_id = thread["user_b_id"] if thread["user_a_id"] == user_id else thread["user_a_id"]
    require_not_blocked(user_id, other_id, "message")

    image_url = (
        upload_image_as_webp("message-images", f"{user_id}/{uuid4()}", image) if image else None
    )

    created = (
        client.table("messages")
        .insert({"thread_id": thread_id, "sender_id": user_id, "body": body, "image_url": image_url})
        .execute()
    )

    is_a = thread["user_a_id"] == user_id
    sender = (thread.get("user_a") if is_a else thread.get("user_b")) or {}
    sender_name = sender.get("display_name") or "Someone"
    other_user_id = thread["user_b_id"] if is_a else thread["user_a_id"]
    if body:
        preview = body if len(body) <= 60 else f"{body[:57]}..."
        message = f'{sender_name} sent you a message: "{preview}"'
    else:
        message = f"{sender_name} sent you a photo"
    notify(other_user_id, "message", message, thread_id=thread_id)

    # A new message un-hides the thread for whoever deleted it, so it isn't lost off their list
    # forever - mirrors `start_thread`'s own revive-on-restart behavior.
    _set_thread_state(client, thread_id, other_user_id, deleted_at=None)

    return created.data[0]


@router.patch("/threads/{thread_id}/mute", response_model=MuteThreadOut)
def set_thread_muted(
    thread_id: str, payload: MuteThreadIn, user_id: str = Depends(get_current_user_id)
) -> dict:
    client = get_supabase_client()
    _require_membership(client, thread_id, user_id)
    _set_thread_state(client, thread_id, user_id, muted=payload.muted)
    return {"id": thread_id, "muted": payload.muted}


@router.delete("/threads/{thread_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_thread(thread_id: str, user_id: str = Depends(get_current_user_id)) -> None:
    """Hides the thread from this user's list only - the other participant's copy (and the
    messages themselves) are untouched, per `message_thread_states`' per-user design."""
    client = get_supabase_client()
    _require_membership(client, thread_id, user_id)
    _set_thread_state(client, thread_id, user_id, deleted_at=datetime.now(UTC).isoformat())
